supervisor/shared/usb/usb_desc.c

```diff
@@ -25,37 +25,308 @@
+    if (usb_midi_enabled()) {
+        total_descriptor_length += usb_midi_descriptor_length();
+    }
+    #endif
+
+    // Now we now how big the configuration descriptor will be, so we can allocate space for it.
+    configuration_descriptor_allocation =
+        allocate_memory(align32_size(total_descriptor_length),
+            /*high_address*/ false, /*movable*/ false);
+    uint8_t *configuration_descriptor = (uint8_t *)configuration_descriptor_allocation->ptr;
+
+    // Copy the template, which is the first part of the descriptor, and fix up its length.
+
+    memcpy(configuration_descriptor, configuration_descriptor_template, sizeof(configuration_descriptor_template));
+
+    configuration_descriptor[CONFIG_TOTAL_LENGTH_LO_INDEX] = total_descriptor_length & 0xFF;
+    configuration_descriptor[CONFIG_TOTAL_LENGTH_HI_INDEX] = (total_descriptor_length >> 8) & 0xFF;
+
+    // Number interfaces and endpoints.
+    // Endpoint 0 is already used for USB control, so start with 1.
+    uint8_t current_interface = 0;
+    uint8_t current_endpoint = 1;
+
+    uint8_t *descriptor_buf_remaining = configuration_descriptor + sizeof(configuration_descriptor_template);
+
+    #if CIRCUITPY_USB_CDC
+    if (usb_cdc_console_enabled()) {
+        // Concatenate and fix up the CDC REPL descriptor.
+        descriptor_buf_remaining += usb_cdc_add_descriptor(
+            descriptor_buf_remaining, &current_interface, &current_endpoint, &current_interface_string, true);
+    }
+    if (usb_cdc_data_enabled()) {
+        // Concatenate and fix up the CDC data descriptor.
+        descriptor_buf_remaining += usb_cdc_add_descriptor(
+            descriptor_buf_remaining, &current_interface, &current_endpoint, &current_interface_string, false);
+    }
+    #endif
+
+    #if CIRCUITPY_USB_MSC
+    if (storage_usb_enabled()) {
+        // Concatenate and fix up the MSC descriptor.
+        descriptor_buf_remaining += storage_usb_add_descriptor(
+            descriptor_buf_remaining, &current_interface, &current_endpoint, &current_interface_string);
+    }
+    #endif
+
+    #if CIRCUITPY_USB_HID
+    if (usb_hid_enabled()) {
+        descriptor_buf_remaining += usb_hid_add_descriptor(
+            descriptor_buf_remaining, &current_interface, &current_endpoint, &current_interface_string,
+            usb_hid_report_descriptor_length());
+    }
+    #endif
+
+    #if CIRCUITPY_USB_MIDI
+    if (usb_midi_enabled()) {
+        // Concatenate and fix up the MIDI descriptor.
+        descriptor_buf_remaining += usb_midi_add_descriptor(
+            descriptor_buf_remaining, &current_interface, &current_endpoint, &current_interface_string);
+    }
+    #endif
+
+    // Now we know how many interfaces have been used.
+    configuration_descriptor[CONFIG_NUM_INTERFACES_INDEX] = current_interface;
+
+    // Did we run out of endpoints?
+    if (current_endpoint - 1 > USB_NUM_EP) {
+        mp_raise_RuntimeError(translate("Not enough USB endpoints"));
```

Comment:
Is this called from within boot.py? I see one call that says it's outside of the VM. Raising an exception will safe mode because the heap isn't active.

Reply:
Changed both `raise`s here to reset into safe mode, with two new safe mode reasons. Also restructured the safe-mode printing slightly to save 32 bytes.